qemu/tests/guest_iommu_group.py

```diff
@@ -0,0 +1,39 @@
+from virttest import error_context
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    this case will:
+    1) Boot guest with virtio devices and iommu is on.
+    2) Check whether the iommu group is seperated correctly.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def verify_iommu_group():
+        """ Verify whether the iommu group is seperated correctly. """
+        error_context.context(
+            'Verify whether the iommu group is seperated correctly.', test.log.info)
+        device_id = session.cmd(
+            "lspci | grep 'PCIe\\|Virtio' | awk '{print $1}'").strip().split()
+        group_id = []
+        i = 0
+        for id in device_id:
+            group_id.append(session.cmd(
+                "dmesg | grep \"iommu group\" | grep '%s' | awk -F \" \" '{print $NF}'"
+                % ("0000:" + id)).strip())
+            test.log.info("Group ID of %s: %s", id, group_id[i])
+            i += 1
+        if len(set(group_id)) != len(group_id):
```

Comment:
When `dmesg | grep "iommu group" | grep '%s' | awk -F " " '{print $NF}'` didn't match, it will return `""`, this means sometimes `len(set(group_id)) == len(group_id)` but it's wrong.

BTW, why check set(grou_id) and group_id, but not device_id?

Reply:
I want to check whether the group_id is duplicate, and make sure that each device is in a different group.